signature/cose/envelope_test.go

```diff
@@ -227,73 +220,679 @@ func TestSignErrors(t *testing.T) {
+		}
+	})
+
+	t.Run("when COSE envelope protected header has malformed expiry", func(t *testing.T) {
+		env, err := getVerifyCOSE("notary.x509", signature.KeyTypeRSA, 3072)
+		if err != nil {
+			t.Fatalf("getVerifyCOSE() failed. Error = %s", err)
+		}
+		env.base.Headers.Protected[headerLabelExpiry] = "malformed"
+		_, err = env.SignerInfo()
+		if err == nil {
+			t.Fatalf("SignerInfo() expects error, but got nil.")
+		}
+	})
+
+	// Testing unprotected headers
+	t.Run("when COSE envelope has malformed certificate chain", func(t *testing.T) {
+		env, err := getVerifyCOSE("notary.x509", signature.KeyTypeRSA, 3072)
+		if err != nil {
+			t.Fatalf("getVerifyCOSE() failed. Error = %s", err)
+		}
+		delete(env.base.Headers.Unprotected, cose.HeaderLabelX5Chain)
+		_, err = env.SignerInfo()
+		if err == nil {
+			t.Fatalf("SignerInfo() expects error, but got nil.")
+		}
+	})
+
+	t.Run("when COSE envelope leaf certificate has wrong type", func(t *testing.T) {
+		env, err := getVerifyCOSE("notary.x509", signature.KeyTypeRSA, 3072)
+		if err != nil {
+			t.Fatalf("getVerifyCOSE() failed. Error = %s", err)
+		}
+		env.base.Headers.Unprotected[cose.HeaderLabelX5Chain] = []interface{}{0}
+		_, err = env.SignerInfo()
+		if err == nil {
+			t.Fatalf("SignerInfo() expects error, but got nil.")
+		}
+	})
+
+	t.Run("when COSE envelope has malformed leaf certificate", func(t *testing.T) {
+		env, err := getVerifyCOSE("notary.x509", signature.KeyTypeRSA, 3072)
+		if err != nil {
+			t.Fatalf("getVerifyCOSE() failed. Error = %s", err)
+		}
+		certs, ok := env.base.Headers.Unprotected[cose.HeaderLabelX5Chain].([]interface{})
+		if !ok || len(certs) == 0 {
+			t.Fatalf("COSE envelope malformed certificate chain")
+		}
+		certRaw, ok := certs[0].([]byte)
+		if !ok {
+			t.Fatalf("COSE envelope malformed leaf certificate")
+		}
+		// Manipulate the leaf certificate
+		certRaw[0] += 'A'
+		certs[0] = certRaw
+		env.base.Headers.Unprotected[cose.HeaderLabelX5Chain] = certs
+		_, err = env.SignerInfo()
+		if err == nil {
+			t.Fatalf("SignerInfo() expects error, but got nil.")
+		}
+	})
+
+}
+
+func TestSignAndVerify(t *testing.T) {
+	env := envelope{}
+	for _, signingScheme := range signingSchemeString {
+		for _, keyType := range keyType {
+			var size []int
+			if keyType == signature.KeyTypeRSA {
+				size = []int{2048, 3072, 4096}
+			} else {
+				size = []int{256, 384, 521}
+			}
+			for _, size := range size {
+				t.Run(fmt.Sprintf("with %s scheme, %v keyType, %v keySize", signingScheme, keyType, size), func(t *testing.T) {
+					// Sign
+					signRequest, err := newSignRequest(signingScheme, keyType, size)
+					if err != nil {
+						t.Fatalf("newSignRequest() failed. Error = %s", err)
+					}
+					encoded, err := env.Sign(signRequest)
+					if err != nil || len(encoded) == 0 {
+						t.Fatalf("Sign() faild. Error = %s", err)
+					}
+
+					// Verify using the same envelope struct
+					_, _, err = env.Verify()
```

Comment:
should we create a new `env` by parsing the `encoded`?

Reply:
This one is intended to use the same envelope as this is the situation when we sign in notation-go. TestSignAndParseVerify does create a new env by UnmarshalCBOR.